cmd/kubelet/app/auth.go

```diff
@@ -95,8 +96,18 @@ func BuildAuthn(client authenticationclient.AuthenticationV1Interface, authn kub
 	}
 
 	return authenticator, func(stopCh <-chan struct{}) {
+		// generate a context from stopCh. This is to avoid modifying files which are relying on this method
+		// TODO: See if we can pass ctx to the current method
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			select {
+			case <-stopCh:
+				cancel() // stopCh closed, so cancel our context
+			case <-ctx.Done():
+			}
```

Comment:
I'm not able to understand well this pattern, does this means that the context can be cancelled from the inside?

Reply:
Yeah, the goal is to have a routine which looks at the stopChannel and cancels the context when stopChannel is closed.

Reply:
> yeah, but AFAIK the context is meant to be cancelled from the parent not from the childs,

yes, this is the recommended use, but here the lifecycle aspect of the context is scoped within the go routine, it's because only when the `stopCh` is closed should we cancel the context. If we return the cancel func to the caller this is no longer guaranteed.

I would recommend:
```
go func() {
                        defer cancel()
			select {
			case <-stopCh:
			case <-ctx.Done():
			}
           }
```

Reply:
i take it back, we don't need a defer in this case.

Reply:
you are only blocking on the stopCh to cancel the context
```
		ctx, cancel := context.WithCancel(context.Background())
		go func() {
			<-stopCh
			cancel() // stopCh closed, so cancel our context
			}
```
the only way this goes out is if someone cancels the context, but that will mind that it has to be done from within the children